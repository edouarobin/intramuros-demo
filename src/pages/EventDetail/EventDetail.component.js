// @flow

import React, { PureComponent } from 'react';
import {
  Image,
  StyleSheet,
  View,
  Text,
  ScrollView,
  StatusBar,
  Dimensions,
  TouchableOpacity,
  Alert,
} from 'react-native';

import {
  GoBackButton,
  EventDetailHeader,
  AddressLabel,
  ContactWeb,
  Schedule,
  Price,
  ImageZoomModal,
  Loader,
  OrganizerCard,
} from 'Intramuros/src/components';
import globalStyle, {
  eventDetailImageHeight,
  STATUS_BAR_HEIGHT,
  BIG_SCREEN_PADDING_OFFSET,
} from 'Intramuros/src/style/globalStyle';
import { HEIGHT } from 'Intramuros/src/components/atoms/EventDetailHeader';
import * as AddCalendarEvent from 'react-native-add-calendar-event';
import { startDateCalendarFormat } from 'Intramuros/src/services/FormatDate';

import { MyFastImage } from 'Intramuros/src/components';
import HTMLText from 'Intramuros/src/utils/HTMLText';
import ENV from 'Intramuros/src/environment';
import { oneEventNormalizer } from '../../redux/events/normalizer';
import request from 'superagent';
import { is } from 'redux-saga/utils';

const screenHeight = Dimensions.get('window').height;

const styles = StyleSheet.create({
  container: {
    backgroundColor: globalStyle.colors.lightGrey,
    flexGrow: 1,
  },
  image: {
    flex: 1,
    height: eventDetailImageHeight,
    //resizeMode: 'contain',
  },
  infoContainer: {
    backgroundColor: 'transparent',
    paddingTop: 20,
    paddingBottom: 0,
    justifyContent: 'flex-start',
    flexDirection: 'column',
    paddingHorizontal: BIG_SCREEN_PADDING_OFFSET,
  },
  descriptionContainer: {
    backgroundColor: 'white',
    flex: 1,
    paddingTop: 0,
    paddingBottom: 16,
    paddingRight: 16 + BIG_SCREEN_PADDING_OFFSET,
    paddingLeft: 16 + BIG_SCREEN_PADDING_OFFSET,
  },
  decriptionStyle: {
    fontFamily: 'Lato-Regular',
    fontSize: 14,
    lineHeight: 24,
    color: 'black',
  },
  authorStyle: {
    fontFamily: 'Lato-Italic',
    marginTop: 16,
    fontSize: 14,
    lineHeight: 24,
    color: 'black',
    // fontStyle: 'italic',
  },
  linksStyle: {
    marginTop: 16,
  },
  moreInfo: {
    fontFamily: 'Lato-BoldItalic',
    // fontStyle: 'italic',
    // fontWeight: 'bold',
    color: globalStyle.colors.greyishBrown,
    marginTop: 10,
    marginBottom: 5,
  },
});

type PropsType = {
  navigation: NavigationPropsType<>,
  eventDetails: EventType,
  isNotification: boolean,
  eventID: number,
};

export default class EventDetail extends PureComponent<PropsType> {
  state = {
    modalVisible: false,
    detailEventFetched: false,
    notifiedEventDetail: null,
  };

  _fetchEvent = (id: number) => {
    try {
      console.log(id);
      request
        .get(`${ENV.API_URL}/events/` + id + `/`)
        .then(response => response.body)
        .then(event => {
          const eventNormalized = oneEventNormalizer(event).entities.event[id];
          this.setState({
            detailEventFetched: true,
            notifiedEventDetail: eventNormalized,
          });
        })
        .catch(error => {
          console.log('ERROR during fetch event from notification : ' + error);
          this.setState({
            detailEventFetched: true,
            fetchEventError: true,
          });
        });
    } catch (error) {
      console.log('Error trying to fetch event detail: ' + error);
      this.setState({
        detailEventFetched: true,
        fetchEventError: true,
      });
    }
  };

  openCalendar = () => {
    const title = this.props.eventDetails.title;
    const location =
      this.props.eventDetails.addressLabel !== title
        ? this.props.eventDetails.addressLabel
        : '';
    const notes = this.props.eventDetails.description;
    const startDate = this.props.eventDetails.startDate + 'T08:00:00.000Z';
    const endDate = this.props.eventDetails.endDate
      ? this.props.eventDetails.endDate + 'T17:00:00.000Z'
      : null;

    var eventConfig = {
      title,
      location,
      notes,
      startDate,
      endDate: endDate ? endDate : startDate,
      allDay: true,
    };

    AddCalendarEvent.presentEventCreatingDialog(eventConfig)
      .then(
        (eventInfo: {
          calendarItemIdentifier: string,
          eventIdentifier: string,
        }) => {
          // handle success - receives an object with `calendarItemIdentifier` and `eventIdentifier` keys, both of type string.
          // These are two different identifiers on iOS.
          // On Android, where they are both equal and represent the event id, also strings.
          // when false is returned, the dialog was dismissed
          if (eventInfo) {
            console.warn(JSON.stringify(eventInfo));
          } else {
            console.warn('dismissed');
          }
        }
      )
      .catch((error: string) => {
        // handle error such as when user rejected permissions
        console.warn(error);
      });
  };

  openZoomImage = () => {
    this.setState({ modalVisible: true });
  };
  closeImageModal = () => {
    this.setState({ modalVisible: false });
  };

  render() {
    let { eventID, isNotification, eventDetails } = this.props;

    if (isNotification && !this.state.detailEventFetched) {
      this._fetchEvent(eventID);
      //Need to wait fetch call ends
      return (
        <View style={{ flex: 1 }}>
          <StatusBar
            barStyle="dark-content"
            backgroundColor="transparent"
            translucent
            animated={true}
          />
          <ScrollView>
            <GoBackButton navigation={this.props.navigation} color="black" />
            <Loader
              isLoading={true}
              message="Chargement en cours"
              fullpage={true}
            />
          </ScrollView>
        </View>
      );
    }

    if (this.state.fetchEventError && !eventDetails) {
      Alert.alert(
        'Oups !',
        "Impossible de récupérer les détails de l'évenement ...",
        [{ text: 'Ok', onPress: () => this.props.navigation.goBack() }],
        {
          cancelable: false,
        }
      );
      return (
        <View style={{ flex: 1 }}>
          <StatusBar
            barStyle="dark-content"
            backgroundColor="transparent"
            translucent
            animated={true}
          />
          <ScrollView>
            <GoBackButton navigation={this.props.navigation} color="black" />
          </ScrollView>
        </View>
      );
    }

    if (isNotification && this.state.notifiedEventDetail) {
      console.log('event from notification call');
      eventDetails = this.state.notifiedEventDetail;
    }

    return (
      <View style={{ flex: 1 }}>
        <StatusBar
          barStyle="dark-content"
          backgroundColor="transparent"
          translucent
          animated={true}
        />
        {eventDetails ? (
          <ScrollView contentContainerStyle={styles.container}>
            <GoBackButton navigation={this.props.navigation} color="black" />
            <View style={styles.infoContainer}>
              <EventDetailHeader
                startDate={eventDetails.startDate}
                endDate={eventDetails.endDate}
                title={eventDetails.title}
                openCalendar={this.openCalendar}
              />

              <View
                style={{
                  marginHorizontal: 20,
                }}
              >
                <AddressLabel
                  address={eventDetails.addressLabel}
                  latitude={eventDetails.latitude}
                  longitude={eventDetails.longitude}
                  label={eventDetails.title}
                />
              </View>
            </View>
            <View style={styles.descriptionContainer}>
              <TouchableOpacity onPress={this.openZoomImage}>
                <MyFastImage
                  style={styles.image}
                  source={{
                    uri: eventDetails.image,
                    priority: 'normal',
                  }}
                  resizeMode={'contain'}
                />
              </TouchableOpacity>
              <HTMLText
                style={styles.decriptionStyle}
                value={eventDetails.description}
                styleContainer={{ marginTop: 16 }}
              />
              {eventDetails.schedule ? (
                <Schedule text={eventDetails.schedule} />
              ) : null}
              {eventDetails.price ? <Price price={eventDetails.price} /> : null}
              {eventDetails.url1 || eventDetails.url2 || eventDetails.url3 ? (
                <View style={styles.linksStyle}>
                  <Text style={styles.moreInfo}>Plus d'informations :</Text>
                  {eventDetails.url1 ? (
                    <ContactWeb
                      website={eventDetails.url1}
                      fisrtElement={true}
                    />
                  ) : null}
                  {eventDetails.url2 ? (
                    <ContactWeb website={eventDetails.url2} />
                  ) : null}
                  {eventDetails.url3 ? (
                    <ContactWeb website={eventDetails.url3} />
                  ) : null}
                </View>
              ) : null}
              {eventDetails.author ? (
                <Text style={styles.authorStyle}>
                  Publié par {eventDetails.author}
                </Text>
              ) : (
                <View />
              )}

              {eventDetails.organizer ? (
                <OrganizerCard
                  title={eventDetails.organizer_name}
                  logo={
                    eventDetails.organizer_logo
                      ? eventDetails.organizer_logo
                      : eventDetails.organizer_category_picto
                  }
                  onPress={() => {
                    if (eventDetails.organizer_type === 'ASSO')
                      this.props.navigation.navigate('AssoDetail', {
                        assoId: eventDetails.organizer,
                        cityOfAsso: eventDetails.organizer_city,
                      });
                    if (eventDetails.organizer_type === 'COMMERCE')
                      this.props.navigation.navigate('CommerceDetail', {
                        commerceId: eventDetails.organizer,
                      });
                  }}
                />
              ) : null}
              <View style={{ height: 60 }} />
            </View>
          </ScrollView>
        ) : (
          <GoBackButton navigation={this.props.navigation} color="black" />
        )}
        {eventDetails && eventDetails.image ? (
          <ImageZoomModal
            image={eventDetails.image}
            modalVisible={this.state.modalVisible}
            closeImageModal={this.closeImageModal}
          />
        ) : null}
      </View>
    );
  }
}
