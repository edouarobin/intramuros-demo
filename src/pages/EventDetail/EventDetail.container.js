// @flow
import { connect } from 'react-redux';

import { selectors } from 'Intramuros/src/redux/events';

import EventDetail from './EventDetail.component';

const mapStateToProps = (rootState: RootStateType, ownProps: any) => ({
  eventDetails: selectors.selectEventDetailsFromID(
    rootState,
    ownProps.navigation.state.params.eventID
  ),
  eventID: ownProps.navigation.state.params.eventID,
  isNotification: ownProps.navigation.state.params.isNotification,
});

const mapDispatchToProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EventDetail);
