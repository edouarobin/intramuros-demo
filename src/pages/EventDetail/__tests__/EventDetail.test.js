// @flow-weak

import React from 'react';
import renderer from 'react-test-renderer';

import EventDetail from '../EventDetail.component';

const mockEventDetails = {
  title: 'title',
  description: 'description',
  author: 'author',
  startDate: '2017-06-12',
  endDate: '2017-06-15',
  city: 2,
  id: 1,
  addressLabel: '48 boulevard des Batignolles, 75013 Paris',
};

it('Fully renders the EventDetails page', () => {
  const tree = renderer
    .create(<EventDetail navigation={{}} eventDetails={mockEventDetails} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
