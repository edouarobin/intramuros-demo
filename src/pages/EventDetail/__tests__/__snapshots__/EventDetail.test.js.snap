// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fully renders the EventDetails page 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <BVLinearGradient
    colors={
      Array [
        3875536895,
        16777215,
      ]
    }
    end={undefined}
    locations={null}
    start={undefined}
    style={
      Object {
        "flex": 1,
        "height": 320.16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingTop": 30,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
            "width": 100,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "width": 100,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 36,
                },
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "black",
                    "marginHorizontal": 8,
                  },
                  Object {
                    "color": "black",
                  },
                ],
                Object {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "black",
                  "fontSize": 16,
                },
                Object {
                  "color": "black",
                },
              ]
            }
          >
            Retour
          </Text>
        </View>
      </View>
    </View>
  </BVLinearGradient>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "backgroundColor": "#F8F8F8",
        "flexGrow": 1,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "uri": undefined,
          }
        }
        style={
          Object {
            "height": 555.6,
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "marginTop": -20,
            "paddingBottom": 14,
            "paddingHorizontal": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFD153",
              "borderRadius": 2,
              "height": 40,
              "justifyContent": "center",
              "marginHorizontal": 8,
              "top": 0,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "black",
                "fontSize": 14,
                "fontWeight": "bold",
              }
            }
          >
            Lundi 12 juin - Jeudi 15 juin
          </Text>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={3}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
              "lineHeight": 28,
              "marginTop": 14,
            }
          }
        >
          title
        </Text>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignSelf": "center",
              "flexDirection": "row",
              "marginBottom": 28,
              "marginTop": 17,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Image
            source="test-file-stub"
            style={
              Object {
                "height": 20,
                "width": 14,
              }
            }
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#2869e5",
                "marginLeft": 10,
              }
            }
          >
            48 boulevard des Batignolles, 75013 Paris
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "white",
            "flex": 1,
            "padding": 16,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "black",
              "fontSize": 14,
              "lineHeight": 24,
            }
          }
        >
          description
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "black",
              "fontSize": 14,
              "fontStyle": "italic",
              "lineHeight": 24,
              "marginTop": 16,
            }
          }
        >
          Publié par 
          author
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
